Add shape metadata to DiskDataset

rbharath opened the pull request:
We've had issues with printing large datasets https://github.com/deepchem/deepchem/issues/1977 since `DiskDataset` doesn't have any way to check the shape of its underlying data without actually iterating through the underlying data on disk. This is really inefficient for large datasets and is also unintuitive (programs will freeze up when you're printing a large dataset).

In this PR, I introduce a strategy to deal with this by adding shape metadata to `DiskDataset`. This stores metadata about shapes in `self.metadata_df` and means that getting shapes for a `DiskDataset` will be very fast and can be done purely in memory. The downside to this is that it induces a change in the storage format for `DiskDataset`. I believe I've made sure that the legacy metadata format can still be constructed and worked with and added some tests for this. I've changed the `__str__`, `__repr__` and `get_shape` functions to rely on this metadata instead of looking at disk.

I think this PR is ready for a first round of review. We'll need to be a little careful on this one since we're changing the tried-and-true metadata for `DiskDataset`. I've added additional documentation to `DiskDataset` making notes about the metadata format and some usage example code to improve maintainability.

CC @peastman @nd-02110114 @ncfrey @seyonechithrananda @vsomnath

rbharath commented:
There's a bit of cruft in here since I started from #2061's code. I'll rebase that out once #2061 is merged in

peastman commented:
Why is the `legacy_metadata` argument needed?  When loading a dataset, we don't need the user to tell us.  We can figure that out just by checking the number of columns in the metadata.  And when creating a new dataset, the only reason to use the old format would be to create a dataset that could be read by older versions of DeepChem, but we already broke forward compatibility by switching from .joblib to .npy files.

peastman commented on deepchem/data/datasets.py:
```diff
@@ -1022,60 +1126,138 @@ def load_metadata(self):
   @staticmethod
-  def _construct_metadata(metadata_entries: List) -> pd.DataFrame:
-    """Construct a dataframe containing metadata.
+  def _save_metadata(tasks, metadata_df, data_dir):
```
```python
   def _save_metadata(tasks: List[str], metadata_df: pd.DataFrame, data_dir: str) -> None:
```

rbharath replied:
Good catch! Will fix

peastman commented on deepchem/data/datasets.py:
```diff
@@ -1092,7 +1274,13 @@ def get_task_names(self) -> np.ndarray:
+
+    Note
+    ----
+    If this `DiskDataset` is in `legacy_metadata` format, reshard will
+    maintain legacy metadata format.
```
Is there any reason not to upgrade it to the new format?  That will give better performance.

rbharath replied:
Good point. I'll upgrade it to the new format

peastman commented on deepchem/data/datasets.py:
```diff
@@ -1473,8 +1665,33 @@ def from_numpy(X: np.ndarray,
+      Tasks in this dataset
+    data_dir: Optional[str], optional (default None)
+      The directory to write this dataset to. If none is specified, will use
+      a temporary dataset instead.
```
I think you meant "a temporary directory instead"?

rbharath replied:
Good catch! Will fix

peastman commented on deepchem/data/datasets.py:
```diff
@@ -957,24 +961,115 @@ def merge(datasets: Sequence[Dataset]) -> "NumpyDataset":
+  `('ids', 'X', 'y', 'w') and not additional shape columns. `DiskDataset`
+  maintains backwards compatibility with this older metadata format, but we
+  recommend for performance reasons not using legacy metadata for new
+  projects.
```
The improved documentation looks great!

rbharath replied:
Thanks!

rbharath commented:
Good point that we can infer whether it's legacy or not! I've gone ahead and removed it from the constructor.

I think it's useful to keep around the option to create datasets in legacy format to facilitate testing for backwards compatibility with reading legacy datasets, so I've left the option in `create_dataset` and `from_numpy`. I've addressed all open comments so I think this is ready for another round of review :)

peastman commented:
I agree with @ncfrey on this, but I'd go even further.  The new format is clearer superior.  There's no reason for a user to want a dataset with the old format.  So why provide an option that users should never actually use?

If we're just concerned about testing, let's stick a small dataset using the old format in the `tests` directory and have a test make sure it still loads correctly.

rbharath commented on deepchem/data/datasets.py:
```diff
@@ -957,24 +961,138 @@ def merge(datasets: Sequence[Dataset]) -> "NumpyDataset":
+  'shard-0-X.npy', 'shard-0-y.npy', etc.
+
+  The basic structure of `DiskDataset` is quite robust and will likely serve
+  you will for datasets up to about 100 GB or larger. However note that
```
Good catch! Will fix

rbharath commented on deepchem/data/datasets.py:
```diff
@@ -1026,30 +1150,66 @@ def load_metadata(self):
 
   @staticmethod
-  def _construct_metadata(metadata_entries: List) -> pd.DataFrame:
+  def _save_metadata(tasks: Optional[Sequence], metadata_df: pd.DataFrame,
+                     data_dir: str) -> None:
+    """Saves the metadata for a DiskDataset
```
Good suggestion! Will do

rbharath commented on deepchem/data/datasets.py:
```diff
@@ -1026,30 +1150,66 @@ def load_metadata(self):
+
+  @staticmethod
+  def _construct_metadata(metadata_entries: List,
+                          legacy_metadata: bool = False) -> pd.DataFrame:
```
Good catch, will fix!

rbharath commented on deepchem/data/datasets.py:
```diff
@@ -957,24 +961,138 @@ def merge(datasets: Sequence[Dataset]) -> "NumpyDataset":
+    if len(self.metadata_df.columns) == 4 and list(
+        self.metadata_df.columns) == ['ids', 'X', 'y', 'w']:
+      logger.info(
+          "Detected legacy metatadata on disk. You can upgrade from legacy metadata to the more efficient current metadata by resharding this dataset."
+      )
```
Good suggestion, will add!

rbharath commented:
@peastman @ncfrey Good points! I've removed support for writing legacy metadata and have added two small test datasets written in the legacy format to the test suite. 

To facilitate testing, I also added a new function `DiskDataset.copy()` which makes a copy of a `DiskDataset` and fixed a bug in `DiskDataset.move`. I've also added tests for moving and copying `DiskDataset`.

rbharath commented:
I've addressed all open comments and only failing tests are the known ones with pytorch-dataset so going to go ahead and merge this in
